raft/src/main/java/org/apache/kafka/raft/UnattachedState.java

```diff
@@ -134,11 +147,45 @@ public boolean canGrantVote(ReplicaKey candidateKey, boolean isLogUpToDate) {
     public String toString() {
         return "Unattached(" +
             "epoch=" + epoch +
+            ", votedKey=" + (votedKey().isPresent() ? votedKey().get() : "null") +
             ", voters=" + voters +
             ", electionTimeoutMs=" + electionTimeoutMs +
+            ", highWatermark=" + highWatermark +
             ')';
     }
 
     @Override
     public void close() {}
+
+    private class VotedState {
+        private final ReplicaKey votedKey;
+
+        VotedState(ReplicaKey votedKey) {
+            this.votedKey = Objects.requireNonNull(votedKey);
+        }
+
+        ReplicaKey votedKey() {
+            return votedKey;
+        }
+
+        ElectionState election() {
+            return ElectionState.withVotedCandidate(epoch, votedKey, voters);
+        }
+
+        boolean canGrantVote(ReplicaKey candidateKey) {
+            if (votedKey.id() == candidateKey.id()) {
+                return !votedKey.directoryId().isPresent() || votedKey.directoryId().equals(candidateKey.directoryId());
+            }
+
+            log.debug(
+                "Rejecting vote request from candidate ({}), already have voted for another " +
+                "candidate ({}) in epoch {}",
+                candidateKey,
+                votedKey,
+                epoch
+            );
+
+            return false;
+        }
```

Comment:
Okay but I am also fine with removing the `VotedState` class and moving this implementation to `UnattachedState::canGrantVote`.

Reply:
I'll opt for removing the subclass then